### PR TITLE
Issue #5102: Fix Java9 build

### PR DESCRIPTION
**Pull request by pbludov**
Issue #5102 

* All references to internal sun.* classes have been removed.
* The doclet is replaced with an internal class with the same functionality but based on the internal parser.

To build Checkstyle with JDK9, all tests and cobertura reports must be disabled.

### CODE REVIEW

**Comment by pbludov** [comment]
Because the javadoc utility converts all the inline tags to their HTML counterparts, I had to do it manually.
Like "{@code &#125;}" => "<code>}</code>". A nice point is that many tags have become easier to read.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 1]
```diff
@@ -480,7 +481,7 @@
     /**
      * An extends clause.  This appear as part of class and interface
      * definitions.  This element appears even if the
-     * {@code extends} keyword is not explicitly used.  The child
+     * <code>extends</code> keyword is not explicitly used.  The child
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: Why do these need to be switched? We just converted these to this type of doc.
See https://github.com/checkstyle/checkstyle/commit/70786709d4f735b9123159a86a752789855af34a#diff-d9a48e7d8e8fa192ed23a1af1597506dR483

File: src/main/java/com/puppycrawl/tools/checkstyle/internal/Main.java [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+ * This class is used internally in the build process to write a property file
+ * with short descriptions (the first sentences) of TokenTypes constants.
+ * Request: 724871
+ * For IDE plugins (like the eclipse plugin) it would be useful to have
+ * a programmatic access to the first sentence of the TokenType constants,
+ * so they can use them in their configuration gui.
+ * @author Pavel Bludov
+ * @noinspection UseOfSystemOutOrSystemErr
+ */
+public final class Main {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/internal/Main.java]: We should not name this `Main` as it will confuse people with our JAR entry `Main`.

File: src/main/java/com/puppycrawl/tools/checkstyle/internal/Main.java [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.internal;
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/internal/Main.java]: Why `internal` and not `doclets`?

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java (hunk 1 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: Because the javadoc utility converts all the inline tags to their HTML counterparts.
If it's ok to have inline tags in the tokentypes.properties file, I'll revert this.

File: src/main/java/com/puppycrawl/tools/checkstyle/internal/Main.java (hunk 3 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/internal/Main.java]: It was easier for me to refactor it in a separate package. To be sure that nothing is missing.
I like the idea of returning all the names as they were.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java (hunk 1 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: I had two options:
* write a converter from inline tags to html tags
* do it manually

"&lt;code&gt;}&lt;/code&gt;" looks better than "{@code &amp;#125;}" in my opinion.

Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: done. Just one additional regexp and one line of code.

File: src/main/java/com/puppycrawl/tools/checkstyle/internal/Main.java (hunk 3 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/internal/Main.java]: done.

File: src/main/java/com/puppycrawl/tools/checkstyle/internal/Main.java (hunk 2 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/internal/Main.java]: done.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 4]
```diff
@@ -3421,7 +3422,7 @@
 
     /**
      * '&amp;' symbol when used in a generic upper or lower bounds constrain
-     * e.g. {@code Comparable&lt;<? extends Serializable, CharSequence>}.
+     * e.g.&nbsp;{@code Comparable&lt;? extends Serializable, CharSequence&gt;}.
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: Why the `&nbsp;`? It is between 2 printable characters so it shouldn't be needed. You usually only need it for multiple spaces in a row.
Shouldn't the brackets be ok since they inside the `{@code` tag?

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 5]
```diff
@@ -3469,7 +3470,7 @@
             GeneratedJavaTokenTypes.BLOCK_COMMENT_BEGIN;
 
     /**
-     * End of block comment: '* /'.
+     * End of block comment: '&#42;/'.
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: Why can't this remain as `*`?

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 6]
```diff
@@ -30,6 +30,7 @@
  *
  * @author Oliver Burn
  * @author <a href="mailto:dobratzp@ele.uri.edu">Peter Dobratz</a>
+ * @noinspection ClassWithTooManyDependents
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: Why is this needed now? We didn't make any changes to this class, except documentation.

File: src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java [hunk 7]
```diff
@@ -19,148 +19,237 @@
...
+     *     or {@code null} if the first sentence is absent or malformed (does not end with period)
+     */
+    private static String getFirstJavadocSentence(DetailAST ast) {
+        String firstSentence = null;
+        final DetailAST modifiers = ast.findFirstToken(TokenTypes.MODIFIERS);
+        for (DetailAST comment = modifiers.getFirstChild(); comment != null;
+                comment = comment.getNextSibling()) {
+            if (comment.getType() == TokenTypes.BLOCK_COMMENT_BEGIN
+                    && JavadocUtils.isJavadocComment(comment)) {
+                final String plainJavadocText = JAVADOC_MULTILINE_TO_SINGLELINE_PATTERN
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java]: If we are going through the hassle of using our Grammar, why do we need a RegExp parser for the javadocs and can't just use our Javadoc grammar?
Regular expressions are known to be bad for javadocs which is why we use the grammar.

File: src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java [hunk 8]
```diff
@@ -19,148 +19,237 @@
...
+            if (commandLine.getArgList().size() == 1) {
+                writePropertiesFile(commandLine.getArgList().get(0),
+                    commandLine.getOptionValue(OPTION_DEST_FILE));
+            }
+            else {
+                printUsage();
+            }
+        }
+        catch (ParseException ex) {
+            System.out.println(ex.getMessage());
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java]: Shouldn't the exception be passed all the way up?
We need a build failure to happen if an exception occurs. No one will be looking at the console for a stack trace unless there is a build failure.

File: src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java [hunk 9]
```diff
@@ -19,148 +19,237 @@
...
+            printUsage();
+        }
+        catch (ANTLRException ex) {
+            ex.printStackTrace();
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java]: same here.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java (hunk 5 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: because this token is '*/' not '* /'.

File: src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java (hunk 7 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java]: The javadoc parser is designed for other purpose. He does not have the tag "first sentence".
Here is what it looks like:
```
    |   |   |   |   `--JAVADOC -> JAVADOC [15:7]
    |   |   |   |       |--NEWLINE -> \n [15:7]
    |   |   |   |       |--LEADING_ASTERISK ->      * [16:0]
    |   |   |   |       |--TEXT ->  '&amp;' symbol when used in a generic upper or lower bounds constrain [16:6]
    |   |   |   |       |--NEWLINE -> \n [16:76]
    |   |   |   |       |--LEADING_ASTERISK ->      * [17:0]
    |   |   |   |       |--TEXT ->  e.g.&nbsp; [17:6]
    |   |   |   |       |--JAVADOC_INLINE_TAG -> JAVADOC_INLINE_TAG [17:17]
    |   |   |   |       |   |--JAVADOC_INLINE_TAG_START -> { [17:17]
    |   |   |   |       |   |--CODE_LITERAL -> @code [17:18]
    |   |   |   |       |   |--WS ->   [17:23]
    |   |   |   |       |   |--TEXT -> Comparable&lt;? extends Serializable, CharSequence&gt; [17:24]
    |   |   |   |       |   `--JAVADOC_INLINE_TAG_END -> } [17:79]
    |   |   |   |       |--TEXT -> . [17:80]
    |   |   |   |       |--NEWLINE -> \n [17:81]
    |   |   |   |       |--TEXT ->       [18:0]
    |   |   |   |       `--EOF -> <EOF> [18:5]
```

We need to do much more then a simple regexp match.
I'll give it a try, but i'm sceptical about it.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java (hunk 5 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: I see now, thanks.
This can be ignored.

File: src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java (hunk 7 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java]: @pbludov See `SummaryJavadocCheck`. It does the same thing and uses our Javadoc parser.

> The javadoc parser is designed for other purpose

The parser is designed for all our Javadoc needs. There is no other purpose for it then that.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java (hunk 4 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: Unfortunately, nothing better can be done. Must be a way to understand where the _first_ sentence ends. For now it is "any period symbol, followed by a space or EOF".
Replacing the nbsp to a space will result the sentence ending in "e.g." which is not correct.

Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: >Shouldn't the brackets be ok since they inside the {@code tag?

They are ok. But please pay attention to the whole file uses &amp;lt;/&amp;gt; notation
```
The {@code &lt;&lt;=} (left shift assignment) operator.
```
No reason to have an exception here. Note also, that there is a typo:
```
Comparable&amp;lt;<
```
both "&amp;lt;" and "<" are present. This line should be fixed anyway.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java (hunk 6 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: Because TokenTypesDoclet depends on TokenTypes.
Any reference to the TokenTypes class leads to the violation "TokenTypes class has too many dependencies (301> 300)".
Roman suggest to suppress it: https://github.com/checkstyle/checkstyle/pull/5312#issuecomment-349306661

File: src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java (hunk 7 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java]: Really? The code you are complain about was taken exactly from this class.
It does not use Javdoc parser.

Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java]: Done. Now the code is more reliable. And slow.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java (hunk 6 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: ok. This is a weird violation where other classes cause a violation in the main class.

File: src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java (hunk 7 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java]: > The code you are complain about was taken exactly from this class.

It extends `AbstractJavadocCheck` which means it does use the Javadoc parser not regular expressions.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/imports/customimportorder/InputCustomImportOrderThirdPartyPackage.java [hunk 10]
```diff
@@ -6,9 +6,9 @@
 import java.*;
 import javax.swing.*;
 
-import static sun.tools.util.CommandLine.parse;
-import static sun.tools.util.ModifierFilter.ALL_ACCESS;
-import static sun.tools.util.ModifierFilter.ALL_ACCESS;
+import static com.puppycrawl.tools.checkstyle.internal.utils.CheckUtil.getTokenText;
+import static com.puppycrawl.tools.checkstyle.internal.utils.CheckUtil.getTokenText;
+import static com.puppycrawl.tools.checkstyle.internal.utils.TestUtil.parseFile;
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/imports/customimportorder/InputCustomImportOrderThirdPartyPackage.java]: Inputs should not rely on our production code. See #4845

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/imports/importorder/InputImportOrder_EclipseDefaultNegative.java [hunk 11]
```diff
@@ -12,7 +12,7 @@
 import javax.swing.JComponent;
 import javax.swing.JTable;
 
-import sun.tools.java.ArrayType;
+import com.puppycrawl.tools.checkstyle.utils.CheckUtils;
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/imports/importorder/InputImportOrder_EclipseDefaultNegative.java]: same.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java (hunk 4 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: > both "&lt;" and "<" are present.

I see the issue now, thanks. This can be ignored.

File: src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java [hunk 12]
```diff
@@ -19,148 +19,288 @@
...
+
+    /**
+     * Parses Java source file. Results in AST which contains comment nodes.
+     * @param file file to parse
+     * @return DetailAST tree
+     * @throws NullPointerException if the text is {@code null}
+     * @throws IOException          if the file could not be read
+     * @throws ANTLRException       if parser or lexer failed
+     */
+    private static DetailAST parseFile(String file) throws IOException, ANTLRException {
```
[the remark this answers is not in the document]
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java]: This will result in a _horizontal_ dependency. It's worse than just copying a couple of lines.

The right solution, IMHO, is to extract this code into an utility class. May be to the TreeWalker itself. I'll take a close look to it.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/imports/customimportorder/InputCustomImportOrderThirdPartyPackage.java (hunk 10 above)
Comment (pbludov) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/imports/customimportorder/InputCustomImportOrderThirdPartyPackage.java]: Thank you! I was looking for something like that, but did not succeeded.
If there is a dedicated issue for import tests, then this PR should be split.

File: src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java [hunk 13]
```diff
@@ -19,148 +19,288 @@
...
+        return firstSentence;
+    }
+
+    /**
+     * Parse block comment DetailAST as Javadoc DetailNode tree.
+     * @param blockComment DetailAST
+     * @return DetailNode tree
+     * @throws CheckstyleException if the javadoc comment can not be parsed
+     */
+    private static DetailNode getDetailNode(DetailAST blockComment) throws CheckstyleException {
```
[the remark this answers is not in the document]
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java]: Done.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/imports/customimportorder/InputCustomImportOrderThirdPartyPackage.java (hunk 10 above)
Comment (pbludov) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/imports/customimportorder/InputCustomImportOrderThirdPartyPackage.java]: Done.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/imports/importorder/InputImportOrder_EclipseDefaultNegative.java (hunk 11 above)
Comment (pbludov) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/imports/importorder/InputImportOrder_EclipseDefaultNegative.java]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/JavadocDetailNodeParser.java [hunk 14]
```diff
@@ -102,7 +103,7 @@
      *        DetailAST of Javadoc comment
      * @return DetailNode tree of Javadoc comment
      */
-    public ParseStatus parseJavadocAsDetailNode(DetailAST javadocCommentAst) {
+    public ParseStatus parseJavadoc(DetailAST javadocCommentAst) {
         blockCommentLineNumber = javadocCommentAst.getLineNo();
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/JavadocDetailNodeParser.java]: This is an evil change. The method name is "parseJavadocAsDetailNode" while return value is "ParseStatus".
I renamed this method to "parseJavadoc" and added a new method named "parseJavadocAsDetailNode" that really returns the DetailNode.
I'm guessing possible problems for plugin developers here.

File: src/main/java/com/puppycrawl/tools/checkstyle/JavadocDetailNodeParser.java [hunk 15]
```diff
@@ -161,6 +162,36 @@ public ParseStatus parseJavadocAsDetailNode(DetailAST javadocCommentAst) {
...
+    /**
+     * Parse block comment DetailAST as Javadoc DetailNode tree.
+     * @param blockComment DetailAST
+     * @return DetailNode tree
+     */
+    public DetailNode parseJavadocAsDetailNode(DetailAST blockComment) {
+        final ParseStatus status = parseJavadoc(blockComment);
+        if (status.getParseErrorMessage() != null) {
+            throw new IllegalArgumentException(getParseErrorMessage(status.getParseErrorMessage()));
+        }
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/JavadocDetailNodeParser.java]: This should be the CheckstyleException. I decided to leave it as it is, maybe there's a reason.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java (hunk 4 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: Unfortunately, it turns out that all encoded symbols inside {@code} remain encoded.
I started a new issue: #5423

**Comment by romani** [comment]
Travis failed items were restarted, @rnveach please finish review.

**Comment by rnveach** [comment]
@pbludov 
> Refactoring of JavadocDetailNodeParser
> Move all parseXXX methods to the Parser class

What do these have to do with Java 9?
Please move refactorings to another PR referencing the PR they are moved to instead of Java 9 issue.

**Comment by romani** [comment]
> JavadocDetailNodeParser.java

should be moved out, as I do not see how it might be related java9.

**Comment by romani** [comment]
>  Issue #5102: TokenTypesDoclet based on internal parser 

I voting to move this update to separate issue, as it is not very related to java9, it should be separate issue, new  issue should reference that it was triggered by java9 update, but after a merge we will still be on java8 (without dependecy to docklets), so it is not java9 fix.

new issue #5431 .

I recommend to close this PR for now. In scope of #5102 , should be changes in build configurations, CIs, anything that actually introduce java9 usage ... .

**Comment by rnveach** [comment]
I assume this PR is on hold until https://github.com/checkstyle/checkstyle/pull/5438 and https://github.com/checkstyle/checkstyle/pull/5437 are done.

**Comment by romani** [comment]
I am closing this PR, as it was split in several, and to ease PR management. Ones we ready for actual java9 changes it is better to open new PR, without history to ease review.
